File: rs/src/model/error_tool.rs
use std::collections::HashMap;

use crate::{__private_types::FieldError, IvoErrorPayload};

#[derive(Debug)]
pub(super) struct ErrorTool<Metadata: Clone> {
    payload: IvoErrorPayload<Metadata>,
}

impl<Metadata: Clone> ErrorTool<Metadata> {
    #[inline(always)]
    pub fn new() -> Self {
        Self {
            payload: HashMap::new(),
        }
    }

    #[inline(always)]
    pub fn set(&mut self, field_name: &str, value: FieldError<Metadata>) {
        self.payload.insert(field_name.to_string(), value);
    }

    #[inline(always)]
    pub fn has_errors(&self) -> bool {
        !self.payload.is_empty()
    }

    #[inline(always)]
    pub fn payload(self) -> IvoErrorPayload<Metadata> {
        self.payload
    }
}

File: ts/src/utils/error-tool.ts
import { sortKeys } from "../utils";
import type { IvoErrorPayload } from "./types";
import type { FieldError } from "./types";

export { ErrorTool };

class ErrorTool<Metadata, PayloadKeys extends string = string> {
  private _payload: IvoErrorPayload<Metadata, PayloadKeys> = {} as never;

  set(field: PayloadKeys, value: FieldError<Metadata>) {
    this._payload[field] = value;
  }

  get fields() {
    return Object.keys(this._payload);
  }

  get hasErrors() {
    return Object.keys(this._payload).length > 0;
  }

  get payload() {
    return sortKeys(this._payload);
  }
}
